src/en/dynasty/src/eu/kanade/tachiyomi/extension/en/dynasty/Dynasty.kt

```diff
@@ -0,0 +1,698 @@
+                    mangas = listOf(entry),
+                    hasNextPage = false,
+                ),
+            )
+        }
+
+        return client.newCall(searchMangaRequest(page, query, filters))
+            .asObservableSuccess()
+            .map { searchMangaParse(it, filters) }
+    }
+
+    override fun getFilterList(): FilterList {
+        val tags = this::class.java
+            .getResourceAsStream("/assets/tags.json")!!
+            .bufferedReader().use { it.readText() }
+            .parseAs<List<Tag>>()
+
+        return FilterList(
+            SortFilter(),
+            TypeFilter(),
+            Filter.Header("Note: Sort and Type may not always work"),
+            Filter.Separator(),
+            TagFilter(tags),
+            AuthorFilter(),
+            ScanlatorFilter(),
+            PairingFilter(),
+            Filter.Header("Note: Author, Scanlator and Pairing filters require exact name. You can add multiple by comma (,) separation"),
+        )
+    }
+
+    // lazy because extension inspector doesn't have implementation
+    private val lruCache by lazy { LruCache<String, Int>(15) }
+
+    override fun searchMangaRequest(page: Int, query: String, filters: FilterList): Request {
+        val typeFilter = filters.firstInstance<TypeFilter>()
+            .also {
+                if (it.checked.isEmpty()) {
+                    throw Exception("Select atleast one type")
+                }
+            }
+
+        val authors = filters.firstInstance<AuthorFilter>().values.map { author ->
+            lruCache[author]
+                ?: fetchTagId(author, "Author")
+                    ?.also { lruCache.put(author, it) }
+                ?: throw Exception("Unknown Author: $author")
+        }
+        val scanlators = filters.firstInstance<ScanlatorFilter>().values.map { scanlator ->
+            lruCache[scanlator]
+                ?: fetchTagId(scanlator, "Scanlator")
+                    ?.also { lruCache.put(scanlator, it) }
+                ?: throw Exception("Unknown Scanlator: $scanlator")
+        }
+        val pairing = filters.firstInstance<PairingFilter>().values.map { pairing ->
+            lruCache[pairing]
+                ?: fetchTagId(pairing, "Pairing")
+                    ?.also { lruCache.put(pairing, it) }
+                ?: throw Exception("Unknown Pairing: $pairing")
+        }
+
+        // series and doujin results are best when chapters are included as type so keep track of this
+        var seriesSelected = false
+        var doujinSelected = false
+        var chapterSelected = false
+
+        val url = "$baseUrl/search".toHttpUrl().newBuilder().apply {
+            addQueryParameter("q", query.trim())
+            filters.firstInstance<SortFilter>().also {
+                addQueryParameter("sort", it.sort)
+            }
+            typeFilter.also {
+                it.checked.forEach { type ->
+                    seriesSelected = seriesSelected || type == "Series"
```

Comment:
String checks are prone to typos. Can you do an instance check here instead?

Reply:
not sure what you mean? `TypeFilter` have property `checked` which returns the current checked values as string

Reply:
I mean if you make a typo within the string the IDE will not help you. While it might be fine now, maintainability wise it's just another possible source of error.
```kt
seriesSelected = seriesSelected || type == "Serie"
```

Reply:
opted for constants